### PR TITLE
Add background filter option to reduce CPU utilization

### PR DESCRIPTION
**Pull request by davcolle**
https://github.com/aws/amazon-chime-sdk-js/compare/blur-cpu-improvement

**Issue #:**

**Description of changes:**

This change adds an option to background blur to enable reducing the frequency of segmenting the foreground from the background. The `filterCPUUtilization` option will set the threshold of percentage of time spent processing background blur. If the percentage of time processing blur exceeds the set threshold, segmentation will be skipped on some frames. This reduction in segmentations will reduce the CPU utilization with the tradeoff of adding a delay to updating the segmentation between the foreground and background.

```javascript
const processor = await BackgroundBlurVideoFrameProcessor.create(null, {filterCPUUtilization: 50});
```

**Testing:**

*Can these tested using a demo application? Please provide reproducible step-by-step instructions.*

Background blur is integrated into the `browser` demo application. To try it out, launch the demo with `npm run start`, join the meeting, click on the camera icon to enable video, then select the filter drop down and select one of the `Background Blur X% CPU` options. You can select 10% - 40% cpu utilization from the drop down.

To manually test changing the `filterCPUUtilization` option go to the `demos/browser/app/meetingV2/meetingV2.ts` file and find the line where `BackgroundBlurVideoFrameProcessor.create` is called and update the code to add the options:

```javascript
this.bbprocessor = await BackgroundBlurVideoFrameProcessor.create(this.getBackgroundBlurSpec(), {filterCPUUtilization: 25});
```

And then run `npm run build` and `npm run start` to run the demo app.

Making this change will cause more skipping of background segmentation. The lower the value of CPU utilization the more likely the blur processor will skip segmentations.

**Checklist:**

1. Have you successfully run `npm run build:release` locally?
Yes

2. Do you add, modify, or delete public API definitions? If yes, has that been reviewed and approved?
Yes. I updated the `BackgroundBlurOptions` interface to add a new option. No existing fields were modified on the interface. This API change still needs to be reviewed.

3. Do you change the wire protocol, e.g. the request method? If yes, has that been reviewed and approved?
No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/backgroundblurprocessor/BackgroundBlurOptions.ts [hunk 1]
```diff
@@ -10,9 +10,19 @@ export default interface BackgroundBlurOptions {
...
-  reportingPeriodMillis?: number;
+  /**
+   * The threshold CPU utilization to trigger skipping background filter updates which will reduce the amount of CPU
+   * used by background filtering. For example, If the reporting period is set to 1000 ms and 500 ms was dedicated to
+   * processing the background filter, then the CPU utilization for that reporting period is 50%. If
+   * {@link filterCPUUtilization} is set to 50 it will cause a `filterCPUUtilizationHigh` event to be fired from the
+   * `BackgroundBlurVideoFrameProcessorObserver`.
+   *
+   */
+  filterCPUUtilization?: number;
```
Comment (dimitarz) [inline src/backgroundblurprocessor/BackgroundBlurOptions.ts:27]: What's the unit here? Percent? I.e 0-100 or 0-1.

File: src/backgroundblurprocessor/BackgroundBlurProcessorProvided.ts [hunk 2]
```diff
@@ -139,15 +146,27 @@ export default class BackgroundBlurProcessorProvided implements BackgroundBlurPr
...
     this.frameCounter = new BackgroundFilterFrameCounter(
       this.delegate,
       options.reportingPeriodMillis,
+      options.filterCPUUtilization,
       this.logger
     );
 
+    this.cpuMonitor = new BackgroundFilterMonitor(options.reportingPeriodMillis * 5, {
+      reduceCPUUtilization: () => {
+        this.updateVideoFramesPerFilterUpdate(Math.min(this.videoFramesPerFilterUpdate + 1, 5));
```
Comment (dimitarz) [inline src/backgroundblurprocessor/BackgroundBlurProcessorProvided.ts:162]: MAX_SKIPPED_SEGMENTATIONS = 5

File: src/backgroundblurprocessor/BackgroundBlurProcessorProvided.ts [hunk 3]
```diff
@@ -139,15 +146,27 @@ export default class BackgroundBlurProcessorProvided implements BackgroundBlurPr
...
     this.setBlurStrength(options.blurStrength);
     this.delegate = new BackgroundBlurVideoFrameProcessorDelegate();
     this.frameCounter = new BackgroundFilterFrameCounter(
       this.delegate,
       options.reportingPeriodMillis,
+      options.filterCPUUtilization,
       this.logger
     );
 
+    this.cpuMonitor = new BackgroundFilterMonitor(options.reportingPeriodMillis * 5, {
```
Comment (dimitarz) [inline src/backgroundblurprocessor/BackgroundBlurProcessorProvided.ts:160]: why the *5?

File: src/backgroundblurprocessor/BackgroundBlurOptions.ts (hunk 1 above)
Comment (davcolle) [reply src/backgroundblurprocessor/BackgroundBlurOptions.ts:27]: I'll update the name and comment to indicate that it is a percentage (0-100)

File: src/backgroundblurprocessor/BackgroundBlurProcessorProvided.ts (hunk 3 above)
Comment (davcolle) [reply src/backgroundblurprocessor/BackgroundBlurProcessorProvided.ts:160]: I wanted to leave some time to adapt. By default the reporting period is 1 second. I think 5 seconds is a good amount of time to settle in to segmentation frequency change and the reevaluate the performance.

File: src/backgroundblurprocessor/BackgroundBlurVideoFrameProcessor.ts [hunk 4]
```diff
@@ -193,6 +193,10 @@ export default class BackgroundBlurVideoFrameProcessor {
     if (!options.reportingPeriodMillis) {
       options.reportingPeriodMillis = 1000;
     }
+
+    if (!options.filterCPUUtilization) {
+      options.filterCPUUtilization = 50;
+    }
```
Comment (dimitarz) [inline src/backgroundblurprocessor/BackgroundBlurVideoFrameProcessor.ts:199]: This is quite high isn't it? On average we see a 20-30% usage. Maybe 30%?

File: src/backgroundblurprocessor/BackgroundBlurProcessorProvided.ts (hunk 3 above)
Comment (dimitarz) [reply src/backgroundblurprocessor/BackgroundBlurProcessorProvided.ts:160]: Ok this can get tricky if somebody sets the reporting period to 5s or 0.5s. Can we just set it to a flat 5s value?

File: src/backgroundblurprocessor/BackgroundBlurProcessorProvided.ts (hunk 2 above)
Comment (ltrung) [inline src/backgroundblurprocessor/BackgroundBlurProcessorProvided.ts:162]: nit: Can we create constant for these value? Not sure what 5 or 1 mean

File: guides/15_Background_Filter_Video_Processor.md [hunk 5]
```diff
@@ -116,9 +116,24 @@ let transformDevice = new DefaultVideoTransformDevice(logger, device, processors
...
+  reportingPeriodMillis?: number;
+  blurStrength?: number;
+  filterCPUUtilization?: number;
+}
+```
+
+* logger: A logger to which log output will be written.
+* reportingPeriodMillis: How often the video frame processor will report observed events. 
+* blurStrength: The amount of blur that will be applied to a video stream. 
+* filterCPUUtilization: The threshold CPU utilization to trigger skipping background filter updates which will reduce the amount of CPU used by background filtering. For example, If the reporting period is set to 1000 ms and 500 ms was dedicated to processing the background filter, then the CPU utilization for that reporting period is 50%. If `filterCPUUtilization` is set to 50 it will cause a `filterCPUUtilizationHigh` event to be fired from the `BackgroundBlurVideoFrameProcessorObserver`.
```
Comment (ltrung) [inline guides/15_Background_Filter_Video_Processor.md:135]: Should builders do anything when receiving `filterCPUUtilizationHigh`?

File: src/backgroundblurprocessor/BackgroundBlurProcessorProvided.ts [hunk 6]
```diff
@@ -415,3 +449,33 @@ export default class BackgroundBlurProcessorProvided implements BackgroundBlurPr
...
+    if (timestamp - this.lastEventTimestamp >= this.monitoringPeriodMillis) {
+      this.lastEventTimestamp = timestamp;
+      this.observer.reduceCPUUtilization();
+    }
+  }
+
+  frameReceived(): void {
+    const timestamp = Date.now();
+    // If a enough time has passed, reset the processor and continue to monitor
+    if (timestamp - this.lastEventTimestamp >= this.monitoringPeriodMillis * 2) {
```
Comment (ltrung) [inline src/backgroundblurprocessor/BackgroundBlurProcessorProvided.ts:476]: So we will always ramp up `videoFramesPerFilterUpdate` after 20s by default regardless of the CPU Utilization at that time?

File: src/backgroundblurprocessor/BackgroundBlurProcessorProvided.ts [hunk 7]
```diff
@@ -303,8 +333,8 @@ export default class BackgroundBlurProcessorProvided implements BackgroundBlurPr
 
       if (this.scaledCanvas === undefined) {
         this.scaledCanvas = document.createElement('canvas');
-        this.scaledCanvas.width = inputCanvas.width * hscale;
-        this.scaledCanvas.height = inputCanvas.height * vscale;
+        this.scaledCanvas.width = this.spec.model.input.width;
```
Comment (ltrung) [inline src/backgroundblurprocessor/BackgroundBlurProcessorProvided.ts:336]: Is this change related? If not, could we document that in CHANGELOG :)

File: src/backgroundblurprocessor/BackgroundBlurProcessorProvided.ts (hunk 2 above)
Comment (davcolle) [reply src/backgroundblurprocessor/BackgroundBlurProcessorProvided.ts:162]: will do.

File: src/backgroundblurprocessor/BackgroundBlurProcessorProvided.ts (hunk 6 above)
Comment (davcolle) [reply src/backgroundblurprocessor/BackgroundBlurProcessorProvided.ts:476]: This will only be called if a high cpu event did not occur in the last 10 seconds. That would mean that the utilization is under the value set in the options and we have some capacity to speed up.

File: src/backgroundblurprocessor/BackgroundBlurProcessorProvided.ts (hunk 7 above)
Comment (davcolle) [reply src/backgroundblurprocessor/BackgroundBlurProcessorProvided.ts:336]: This is unrelated. I will update in changelog.

File: guides/15_Background_Filter_Video_Processor.md (hunk 5 above)
Comment (davcolle) [reply guides/15_Background_Filter_Video_Processor.md:135]: I will add an example implementation and what action to take if this fires. Mainly you can disable background blur if the utilization is too high.

File: src/backgroundblurprocessor/BackgroundBlurProcessorProvided.ts (hunk 6 above)
Comment (ltrung) [reply src/backgroundblurprocessor/BackgroundBlurProcessorProvided.ts:476]: Thanks. Also can we add it to the demo also?

File: guides/15_Background_Filter_Video_Processor.md [hunk 8]
```diff
@@ -116,21 +116,40 @@ let transformDevice = new DefaultVideoTransformDevice(logger, device, processors
...
+
 ## Observer notifications
 
 You can optionally implement the `BackgroundBlurVideoFrameProcessorObserver` interface and use `addObserver` to receive callbacks when an event occurs:
 
 * `filterFrameDurationHigh`:  This event occurs when the amount of time it takes to apply the background blur is longer than expected. The measurement is taken from the time the process method starts to when it returns. For example, if the video is running at 15 frames per seconds and we are averaging more than 67 ms (1000 ms reporting period / 15 fps) to apply background blur, then a very large portion of each frame's maximum processing time is taken up by processing background blur.
+* `filterCPUUtilizationHigh`: This event occurs when the CPU utilization of background filtering exceeds the `filterCPUUtilization` defined in the `BackgroundBlurOptions`. This can be used as a trigger to disable background blur if CPU utilization is too high.
```
Comment (ltrung) [inline guides/15_Background_Filter_Video_Processor.md:142]: Should we document that JS SDK also attempts to migrate this?

Comment (davcolle) [reply guides/15_Background_Filter_Video_Processor.md:142]: yes. I'll update.

File: src/backgroundblurprocessor/BackgroundBlurProcessorProvided.ts (hunk 6 above)
Comment (ltrung) [reply src/backgroundblurprocessor/BackgroundBlurProcessorProvided.ts:484]: Can we rename the variable `lastEventTimestamp` to indicate that it is the last timestamp from high cpu utilization event?

**Review by ltrung — DISMISSED** [review]

**Comment by xuesichao** [comment]
@davcolle Could you please resolve the conflict?

**Review by ltrung — APPROVED** [review]
